File: crates/mithril-oxide-sys/cpp/Dialect/Async/IR/AsyncTypes.cpp
#include "AsyncTypes.hpp"

#include <mlir/Dialect/Async/IR/AsyncTypes.h>
#include <mlir/IR/MLIRContext.h>
#include <mlir/IR/Types.h>


namespace mithril_oxide_sys::async {

const void *CoroHandleType_get(MLIRContext *ctx)
{
    return CoroHandleType::get(ctx).getAsOpaquePointer();
}

const void *CoroIdType_get(MLIRContext *ctx)
{
    return CoroIdType::get(ctx).getAsOpaquePointer();
}

const void *CoroStateType_get(MLIRContext *ctx)
{
    return CoroStateType::get(ctx).getAsOpaquePointer();
}

const void *GroupType_get(MLIRContext *ctx)
{
    return GroupType::get(ctx).getAsOpaquePointer();
}

const void *TokenType_get(MLIRContext *ctx)
{
    return TokenType::get(ctx).getAsOpaquePointer();
}

const void *ValueType_get(const void *type)
{
    return ValueType::get(mlir::Type::getFromOpaquePointer(type)).getAsOpaquePointer();
}

} // namespace mithril_oxide_sys::async

File: crates/mithril-oxide-sys/src/Dialect/Async/IR/AsyncTypes.rs
#[cxx::bridge]
mod ffi {
    #[namespace = "mlir"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/Async/IR/AsyncTypes.hpp");

        type MLIRContext = crate::IR::MLIRContext::MLIRContext;
    }

    #[namespace = "mithril_oxide_sys::async"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/Async/IR/AsyncTypes.hpp");

        type c_void = crate::IR::Value::ffi::c_void;

        unsafe fn CoroHandleType_get(context: *mut MLIRContext) -> *const c_void;
        unsafe fn CoroIdType_get(context: *mut MLIRContext) -> *const c_void;
        unsafe fn CoroStateType_get(context: *mut MLIRContext) -> *const c_void;
        unsafe fn GroupType_get(context: *mut MLIRContext) -> *const c_void;
        unsafe fn TokenType_get(context: *mut MLIRContext) -> *const c_void;
        unsafe fn ValueType_get(type_: *const c_void) -> *const c_void;
    }
}
